### DISCUSSION TITLE
How can I get the path of Android internal storage in Dioxus@0.6?

### DISCUSSION THREAD

**Discussion by imbolc** [DioxusLabs/dioxus#3475] [Q&A] [answered]
I'm guessing I should use `android_activity::AndroidApp::internal_data_path`. But how can I access access to `AndroidApp` in Dioxus?

Crosslinks:

- [Rust forum](https://users.rust-lang.org/t/how-can-i-get-the-path-of-android-internal-storage-in-dioxus-0-6/123291)
- [StackOverflow](https://stackoverflow.com/questions/79319500/how-can-i-get-the-path-of-android-internal-storage-in-dioxus0-6)
- [Reddit](https://www.reddit.com/r/rust/comments/1hqahqg/how_can_i_get_the_path_of_android_internal/)

**Comment by LaurentMazare** [comment, marked as answer]
I came across a similar issue today (finding some local directory where to store some persistent files) and came up with something like this. Disclaimer: I'm a total newbie when it comes to android development so I don't have much of a clue whether this is reasonable or not and how it compares to using `internal_data_path`.
```rust
fn get_cache_dir() -> Result<String> {
    let ctx = ndk_context::android_context();
    let vm = unsafe { jni::JavaVM::from_raw(ctx.vm().cast()) }?;
    let mut env = vm.attach_current_thread()?;
    let ctx = unsafe { jni::objects::JObject::from_raw(ctx.context().cast()) };
    let cache_dir = env
        .call_method(ctx, "getFilesDir", "()Ljava/io/File;", &[])?
        .l()?;
    let cache_dir: jni::objects::JString = env
        .call_method(&cache_dir, "toString", "()Ljava/lang/String;", &[])?
        .l()?
        .try_into()?;
    let cache_dir = env.get_string(&cache_dir)?;
    let cache_dir = cache_dir.to_str()?;
    Ok(cache_dir.to_string())
}
```

**Reply by imbolc** [reply]
Seems working, thanks a lot :pray:

**Reply by jkelleyrtp** [reply]
I'd like to expose the android activity in the public API! Just hasn't been done yet.

**Comment by ymelois** [comment]
I had the same issue, here's my solution. I used `wry::prelude::dispatch` to get the env and app's activity without unsafe calls. It works the same way as tauri does: https://github.com/tauri-apps/tauri/blob/231e9a5ee1ef5094093ea526db80caf2ccfcfe03/crates/tauri/src/plugin/mobile.rs#L395

```rust
#[cfg(target_os = "android")]
fn app_local_data_dir() -> Result<PathBuf> {
    use jni::objects::{JObject, JString};
    use jni::JNIEnv;

    let (tx, rx) = std::sync::mpsc::channel();

    fn run(env: &mut JNIEnv<'_>, activity: &JObject<'_>) -> Result<PathBuf> {
        let files_dir = env
            .call_method(activity, "getFilesDir", "()Ljava/io/File;", &[])?
            .l()?;
        let files_dir: JString<'_> = env
            .call_method(files_dir, "getAbsolutePath", "()Ljava/lang/String;", &[])?
            .l()?
            .into();
        let files_dir: String = env.get_string(&files_dir)?.into();
        Ok(PathBuf::from(files_dir))
    }

    wry::prelude::dispatch(move |env, activity, _webview| tx.send(run(env, activity)).unwrap());

    rx.recv().unwrap()
}
```

**Reply by imbolc** [reply]
For some reason, when I swap @LaurentMazare's code with this, the app closes in the emulator, and I don’t know how to debug the issue.

**Reply by ymelois** [reply]
Did you use wry from dioxus::mobile::wry or another version? I got the same behavior of the app closing instantly when not using the wry re-export (or simply not the same version)

**Reply by imbolc** [reply]
Works with `dioxus::mobile::wry`, thank you :pray:
